### PR TITLE
[Spike] Demo next/prev story functionality

### PR DESCRIPTION
**Pull request by mwickett**
Quick and dirty spike to try out an approach for handling next / previous links in the story view.

* Generates a JSON array of story IDs
* Uses that array in `getStaticProps` to find the left and right story IDs of the current story

This is meant to stir conversation, lots of tidying up and hardening to be done if we go forward.

### CODE REVIEW

**Comment by mwickett** [comment]
@LuckeeDev Tagging you on this as well for thoughts!

**Comment by silent1mezzo** [comment]
#### 📝Changed routes:


| [**/story/\[id\]**](https://www-llspsh3pl-my-covid-story.vercel.app/story/[id]) |
|:-:|
|<a href="https://www-llspsh3pl-my-covid-story.vercel.app/story/[id]">Screenshot of /story/[id]</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/2DZkamlSG0X-754j2QbczFmIW_qxWp7u3ESf8SWli7hB7RJsAoVnn6vGZxNnzITs2etGnABxCji1RZpXrM8nnQ.png">(view full size)</a> |







Commit e3b10dbe7527be350f88e4611fd4bb45d08a36b1 (https://www-llspsh3pl-my-covid-story.vercel.app).

**Comment by mwickett** [comment]
Ah, gotta tidy up the build command to make this work. Coming back to this later today!

File: generated/story-ids.json [hunk 1]
```diff
@@ -0,0 +1,91 @@
+[
```
Comment (LuckeeDev) [inline generated/story-ids.json:1]: Are we sure we want to commit to the repo a statically generated file?

File: pages/story/[id].tsx [hunk 2]
```diff
@@ -87,6 +95,10 @@ export default function StoryPage(props: StoryPageProps): JSX.Element {
 
       <Box>
         <StoryDetail story={story} onShare={onOpen} />
+        <Box>
+          {story.previous && <Link href={`${story.previous}`}>Previous Story</Link>}
+          {story.next && <Link href={`${story.next}`}>Next Story</Link>}
```
Comment (LuckeeDev) [inline pages/story/[id].tsx:100]: This can be simplified into `href={story.previous}` and `href={story.next}`.

**Review by LuckeeDev — COMMENTED** [review]
I'm currently a little busy but this is what I found looking quickly through the code! I'll definitely dig in a little deeper later today/tomorrow.

File: generated/story-ids.json (hunk 1 above)
Comment (mwickett) [reply generated/story-ids.json:1]: yeah, I need to gitignore this! (This is just a demo PR, so it's messy)

**Comment by JamiesonRoberts** [comment]
@mwickett Concerns/Thoughts:

1. This is that it puts us right back into the place where the file will only be generated on deploy, so it wont be dynamic on the site.
2. How does this function if a user has filtered before hitting the individual story? Does this mean that we need to keep our story list larger with more data?
3. Why not have the next/previous URL's be resolved on click rather than statically on page load. Shift the burden entirely client side so that we can pass in whatever we want to determine the next one? This would leave only one edge case which is if the story has no step in either direction, but that could be solved by then sending the user back to the home page

**Comment by LuckeeDev** [comment]
> 3. Why not have the next/previous URL's be resolved on click rather than statically on page load. Shift the burden entirely client side so that we can pass in whatever we want to determine the next one?

This seems like the better option to me. I think it's also more future-proof since we might later add advanced custom filtering, sorting, searching and maybe even random story suggestions, which all require a flexible client side solution instead of a statically generated one.

How would you think we'd pull this off @JamiesonRoberts?

I have some ideas, will be able to share more later.

P.S: we might also consider adding a "similar stories" section in addition to the next/prev feature.

**Comment by LuckeeDev** [comment]
### Concerns
I'm afraid statically generating a JSON file is not the way to go. When will it be generated? Only at build time? And when will it be available? Only for the first render or for all revalidations?

### Idea
I think that, as the app scales, we will need some sort of state management solution. For the sake of simplicity, I thought of a mix of Context API and `useReducer()` instead of a full suite like Redux that would probably be overly complex for what we're trying to achieve. Thinking of state management, I figured that this would be a perfect use case for it. Here's what I would do and how it links to @mwickett's work.

- We can set up two global Contexts in `_app.tsx` for retrieving the state and dispatching to a `useReducer()` function which also lives at the app level.
- Since in `index.tsx` we have access to the full collection of stories, we can dispatch an action on the first render (inside a `useEffect()` hook) to make this list available to the whole app through the Context.
- With this data, we can then do all sorts of crazy stuff as @JamiesonRoberts suggested and have a very flexible next/prev feature.

There's only an issue with this approach: what if the user first hits a story page (let's say, coming from Twitter) and then the index? I thought about it and there were only two solutions that came to my mind: we could statically generate a fallback version of the next/prev feature for this specific edge case or, instead, remove the next/prev feature for this specific edge case (this seems like the hacky and not optimal way of solving it).

### Conclusion
I hope I wrote it out clearly enough, I'd love to hear your thoughts on this guys! @mwickett @JamiesonRoberts @silent1mezzo

**Comment by davesteinberg** [comment]
I also favour a client side solution using context (probably with a static, sever-side fallback for when directly open a story link, bypassing the feed). I've got it all in my head, and partly sketched out, but I've just been interrupted by other things...I hope to be able to share it soon...

**Pull request closed without merging**